### PR TITLE
#422 BUG: Pyspark pipeline generates duplicated data records under pipeline with combine records

### PR DESCRIPTION
**Pull request by jaebchoi**
Issue 422 https://github.com/boozallen/aissemble/issues/422

### CODE REVIEW

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java [hunk 1]
```diff
@@ -36,6 +37,7 @@ public static VelocityContext populateCommonPythonContext(VelocityContext veloci
         if (velocityContext == null) {
             velocityContext = new VelocityContext();
         }
+        rootArtifactId = generationContext.getRootArtifactId();
```
Comment (jaebchoi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java:40]: I: This is done so that we can retrieve rootArtifactID in the metamodel package.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 2]
```diff
@@ -135,8 +136,9 @@ private Set<String> getImportsWithRelativePaths(boolean isImplModule) {
         for (String moduleImport : imports) {
             if (moduleImport.startsWith("from record.")
                     || moduleImport.startsWith("from dictionary.")) {
+                String rootArtifact = PythonGeneratorUtils.rootArtifactId.replace("-", "_");
                 importsWithRelativePaths.add(moduleImport.replace("from ",
-                        isImplModule ? "from .." : "from ..."));
+                        "from " + rootArtifact + "_data_records."));
```
Comment (jaebchoi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:141]: I: few of python imports had relative imports, because they did generated data records under pyspark directory.

Now that we deleted duplicated data record in pyspark, needed to point those into correct package.

File: foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm [hunk 3]
```diff
@@ -24,7 +24,7 @@
                 <artifactId>fermenter-mda</artifactId>
                 <configuration>
                     <basePackage>${moduleArtifactIdPythonCase}</basePackage>
-                    <profile>data-delivery-pyspark</profile>
+                    <profile>data-delivery-pyspark-pipeline</profile>
                     <propertyVariables>
```
Comment (meliz19) [inline foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm:28]: What prompted this change?

Comment (jaebchoi) [reply foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm:28]: This is actually core fix to remove duplication. According to profile.json, data-delivery-pyspark profile had unnecessary profile references which led to duplication. Just importing data-delivery-pyspark-pipeline can be sufficient.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java [hunk 4]
```diff
@@ -20,6 +20,7 @@
  * across {@link org.technologybrewery.fermenter.mda.generator.Generator} hierarchies.
  */
 public final class PythonGeneratorUtils {
+    public static String rootArtifactId;
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java:23]: S: This shared state seems like a bad idea, and doesn't seem cohesive with the class goals at all.  I think we're probably missing some parameter passing somewhere.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 5]
```diff
@@ -135,8 +136,9 @@ private Set<String> getImportsWithRelativePaths(boolean isImplModule) {
         for (String moduleImport : imports) {
             if (moduleImport.startsWith("from record.")
                     || moduleImport.startsWith("from dictionary.")) {
+                String rootArtifact = PythonGeneratorUtils.rootArtifactId.replace("-", "_");
                 importsWithRelativePaths.add(moduleImport.replace("from ",
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:140]: A: The name of this method/variable seems misleading now. These aren't relative imports.

Comment (jaebchoi) [reply foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:140]: yup now it isrelative imports anymore, Ill update to something like getDataRecordsAndDictImports.

File: foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm [hunk 6]
```diff
@@ -24,7 +24,7 @@
                 <artifactId>fermenter-mda</artifactId>
                 <configuration>
                     <basePackage>${moduleArtifactIdPythonCase}</basePackage>
-                    <profile>data-delivery-pyspark</profile>
+                    <profile>data-delivery-pyspark-pipeline</profile>
```
[the remark this answers is not in the document]
Comment (jaebchoi) [reply foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm:27]: Ill test to make sure, didn't realize this use case!

[the remark this answers is not in the document]
Comment (jaebchoi) [reply foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm:27]: Okay so old monolith generation did failed with my new impl, so I kept the old implementation to support both cases based on profile name of pipleline

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java (hunk 4 above)
Comment (jaebchoi) [reply foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/PythonGeneratorUtils.java]: updated to pass in artifactID instead of setting public static variables.

File: foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm (hunk 6 above)
[the remark this answers is not in the document]
Comment (jaebchoi) [reply foundation/foundation-mda/src/main/resources/templates/data-delivery-pyspark/pipeline.pom.xml.vm:27]: Verified old monolith gen works

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java [hunk 7]
```diff
@@ -214,6 +214,45 @@ public String getPackaging() {
             return packaging;
         }
     }
+
```
Comment (jaebchoi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java:217]: I: This is needed so that I can retrieve profile Name from maven pom so that I can handle Old Monolith Generation and new shared data records use cases.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 8]
```diff
@@ -130,18 +133,25 @@ public Set<String> getImplImports() {
      * @return {@link #imports} that have been appropriately modified to include the correct
      * relative import paths for any references to metamodel-define records or dictionary types.
      */
-    private Set<String> getImportsWithRelativePaths(boolean isImplModule) {
-        Set<String> importsWithRelativePaths = new TreeSet<>();
+    private Set<String> getImports(boolean isImplModule) {
```
Comment (jaebchoi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:136]: I: this is to retrieve all imports including records and dict so it would make sense to name it as just getImports.

File: docs/modules/ROOT/pages/semantic-data.adoc [hunk 9]
```diff
@@ -220,7 +220,8 @@ Additionally, update any `import` statements in your pipeline code to reflect th
 
 In some very targeted cases where the data model will never need to be re-used and there is only one pipeline that
 processes the data, the semantic data classes can be directly generated into the pipeline module. In general, this is
-the legacy behavior of aiSSEMBLE and should only be undertaken with due consideration. In order to do this, update the
+the legacy behavior of aiSSEMBLE and should only be undertaken with due consideration. In order to do this, first delete
+generated spark pipelines if necessary and update the
```
Comment (jaebchoi) [inline docs/modules/ROOT/pages/semantic-data.adoc:224]: I: I added this because mvn clean build doesn't overwrite already generated files.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java [hunk 10]
```diff
@@ -214,6 +214,45 @@ public String getPackaging() {
...
+            for (int i = 0; i < nodeList.getLength(); i++){
+                Node node = nodeList.item(i);
+
+                if (node.getNodeType() == Node.ELEMENT_NODE){
+
+                    Element element = (Element) node;
+                    String field = element.getTextContent();
+                    if(field.startsWith("data-delivery"))
+                    {
+                        return field;
```
Comment (csun-cpointe) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java:246]: Q: it seems we always return the first profile name starting with `data-delivery`, is it possible that we have multiple profile name starting with `data-delivery` and can we be more specific what profile name we are looking for?

Comment (jaebchoi) [reply foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java:246]: yup I missed that, this could cause issues when there're multiple profile that starts data-delivery by looking at the profiles.json
What I am seeking for is whether pom has data-delivery-pyspark or data-delivery-pyspark.
Upon investigating profiles.json, we have only two profiles that starts with data-delivery-pyspark so Ill update to start with data-delivery-pyspark.

In some rare edge case we add data-delivery-pyspark and data-delivery-pyspark-pipeline in a pom file, that's okay. There will be many duplication in code but we have cases that supports both, so just getting first profile starting data-delivery-pyspark is good.

**Review by csun-cpointe — APPROVED** [review]

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java [hunk 11]
```diff
@@ -214,6 +214,44 @@ public String getPackaging() {
...
+                    String field = element.getTextContent();
+                    if(field.equals("data-delivery-pyspark") || field.equals("data-delivery-pyspark-pipeline") )
+                    {
+                        return field;
+                    }
+                }
+            }
+        } catch (Exception e){
+            logger.error("Unable to find profileName", e);
+            throw new NoSuchElementException("No profile name identified within " + pomFile.getPath());
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java:251]: A: We should not both log an exception and throw. That results in double-logging and cluttered log files that are hard to parse. We should just re-throw the exception in a wrapper that gives a specific message, and let higher code catch and log as appropriate. E.g.: `throw new FermenterException("Failed to locate pipeline generation Fermenter profile in " + pomFile.getPath(), e);`

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java [hunk 12]
```diff
@@ -214,6 +214,44 @@ public String getPackaging() {
...
+        }
+        return null;
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/generator/util/MavenUtil.java:253]: A: I don't think I agree with returning `null` here.  We cannot perform generation properly if we don't know whether data records are being included in the same source. I'd even prefer this method maybe return a boolean and be called `isRecordGenerationInPipelineModule`. Then return false if you don't find `data-delivery-pyspark` exactly. This would keep the code more flexible in case we add more/different pipeline-specific profiles in the future.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 13]
```diff
@@ -130,18 +135,25 @@ public Set<String> getImplImports() {
...
+        Set<String> importsSet = new TreeSet<>();
         for (String moduleImport : imports) {
             if (moduleImport.startsWith("from record.")
                     || moduleImport.startsWith("from dictionary.")) {
-                importsWithRelativePaths.add(moduleImport.replace("from ",
-                        isImplModule ? "from .." : "from ..."));
+                if(profileName.equals("data-delivery-pyspark"))
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:143]: A: Given we allow `profileName` to be `null` this is susceptible to a hard to trace NPE. In general, a good practice is to put the constant on the LHS of the equals just in case. E.g. `"data-delivery-pyspark".equals(profileName)`

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 14]
```diff
@@ -130,18 +135,25 @@ public Set<String> getImplImports() {
...
+                {
+                    importsSet.add(moduleImport.replace("from ",
+                            isImplModule ? "from .." : "from ..."));
+                }else if(profileName.equals("data-delivery-pyspark-pipeline")){
+                    String rootArtifact = rootArtifactId.replace("-", "_");
+                    importsSet.add(moduleImport.replace("from ",
+                            "from " + rootArtifact + "_data_records."));
+                }
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:151]: A: else? We could consider throwing an exception in the final else block, again just to fail faster and help track down bugs by signaling they're true root faster.

File: foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java [hunk 15]
```diff
@@ -130,18 +135,25 @@ public Set<String> getImplImports() {
...
             if (moduleImport.startsWith("from record.")
                     || moduleImport.startsWith("from dictionary.")) {
-                importsWithRelativePaths.add(moduleImport.replace("from ",
-                        isImplModule ? "from .." : "from ..."));
+                if(profileName.equals("data-delivery-pyspark"))
+                {
+                    importsSet.add(moduleImport.replace("from ",
+                            isImplModule ? "from .." : "from ..."));
+                }else if(profileName.equals("data-delivery-pyspark-pipeline")){
+                    String rootArtifact = rootArtifactId.replace("-", "_");
```
Comment (ewilkins-csi) [inline foundation/foundation-mda/src/main/java/com/boozallen/aiops/mda/metamodel/element/python/PythonStep.java:148]: A: I think there is a util method that will convert an artifact ID to a python package name.
